File: config/core/schemas/crop.py
"""
KisanAI OS
Crop Schemas
Version: 5.2.0
"""

from typing import Optional

from pydantic import BaseModel, Field, field_validator


def _clean_required(value):
    """Strip + collapse whitespace; reject blank input."""
    if value is None:
        raise ValueError("value is required")

    text = " ".join(str(value).strip().split())

    if not text:
        raise ValueError("value cannot be blank")

    return text


class CropCreate(BaseModel):
    crop_id: Optional[int] = None
    farmer_id: Optional[int] = None
    crop_name: str
    season: str
    duration_days: int = Field(gt=0)
    water_requirement: str

    @field_validator("crop_name", "season", "water_requirement", mode="before")
    @classmethod
    def _normalize_text(cls, value):
        return _clean_required(value)


class CropUpdate(BaseModel):
    farmer_id: Optional[int] = None
    crop_name: str
    season: str
    duration_days: int = Field(gt=0)
    water_requirement: str

    @field_validator("crop_name", "season", "water_requirement", mode="before")
    @classmethod
    def _normalize_text(cls, value):
        return _clean_required(value)


class CropOut(BaseModel):
    crop_id: int
    farmer_id: Optional[int] = None
    crop_name: str
    season: str
    duration_days: int
    water_requirement: str
    created_at: str

File: mobile/lib/models/crop.dart
/// Crop model (`CropOut`).
class Crop {
  const Crop({
    required this.cropId,
    required this.farmerId,
    required this.cropName,
    required this.season,
    required this.durationDays,
    required this.waterRequirement,
    required this.createdAt,
  });

  final int cropId;
  final int? farmerId;
  final String cropName;
  final String season;
  final int durationDays;
  final String waterRequirement;
  final String createdAt;

  factory Crop.fromJson(Object? json) {
    final map = json as Map<String, dynamic>;
    return Crop(
      cropId: (map['crop_id'] as num).toInt(),
      farmerId: (map['farmer_id'] as num?)?.toInt(),
      cropName: map['crop_name'] as String? ?? '',
      season: map['season'] as String? ?? '',
      durationDays: (map['duration_days'] as num?)?.toInt() ?? 0,
      waterRequirement: map['water_requirement'] as String? ?? '',
      createdAt: map['created_at'] as String? ?? '',
    );
  }
}
